Thread-safe version of sparse function in cv::gpu::PyrLKOpticalFlow

juanecito opened the pull request:
Issue: The function sparse of cv::gpu::PyrLKOpticalFlow class isn't thread-safe.

Resolution: This modification permits to call this function from different CPU threads and to use CUDA texture resources in 5 parallel execution lines. Critical section is also managed in this function (TBB intel*). In the future a dense function thread-safe version can be made.

Best regards
Juan María Gómez López

*Fixed: (first version used posix threads)

```
force_builders=Custom
docker_image:Custom=ubuntu-cuda:16.04
buildworker:Linux x64=linux-1
buildworker:Linux x64 Debug=linux-1
```

alalek commented:
2.4 branch is in a deep maintenance mode, only bugfixes are accepted.

> pthread_mutex_t

OpenCV is cross-platform library, so this code breaks builds on Windows.

alalek commented on modules/gpu/src/cuda/pyrlk.cu:
```diff
@@ -309,6 +402,198 @@ namespace pyrlk
+
+        float D = A11 * A22 - A12 * A12;
+
+        if (D < numeric_limits<float>::epsilon())
```
Such checks should have this form: `(fabs(D) < ...)`

alalek commented on modules/gpu/src/pyrlk.cpp:
```diff
@@ -40,6 +40,8 @@
 
+#include <pthread.h>
+
 #include "precomp.hpp"
```
"precomp.hpp" must be the first include in .cpp file.

alalek commented on modules/gpu/src/pyrlk.cpp:
```diff
@@ -98,7 +107,9 @@ namespace
-void cv::gpu::PyrLKOpticalFlow::sparse(const GpuMat& prevImg, const GpuMat& nextImg, const GpuMat& prevPts, GpuMat& nextPts, GpuMat& status, GpuMat* err)
+void cv::gpu::PyrLKOpticalFlow::sparse(const GpuMat& prevImg,
+						const GpuMat& nextImg, const GpuMat& prevPts,
+						GpuMat& nextPts, GpuMat& status, GpuMat* err)
```
Don't use "tabs"

juanecito replied:
Ok, I'm going to change pthreads by c++11 std::mutex and std::condition_variable and add accuracy tests and performance tests

alalek replied:
C++11 is not allowed for OpenCV 2.4/3.x - C++98 is required.

juanecito replied:
I understand, so I can use TBB (tbb::mutex and condition_variable), and make compilable the thread-safe version only if there is TBB support
"#ifdef HAVE_TBB"
What do you think?
Thank you for your help

juanecito commented:
In second commit i have fixed tabs and pthread, and I have added performance and accuracy test

juanecito commented:
Hi alalek, should I do anything more to complete this pull-request?, is there anything to fix?

vpisarev commented:
@juanecito, sorry for delay with reply. As you can see, your pull request does not pass our buildbot, there are compile errors (see the right-most 'Custom' builder).

juanecito commented:
I think HAVE_TBB macro isn't defined in .cu files (missing in nvcc compiler flags). I've configured cmake  with -D WITH_TBB=ON option but the functions which are defined inside conditional-compilation block HAVE_TBB aren't compilated. I've added a fix including cvconfig.h in pyrlk.cu in another pullrequest (#11534 ). On the other hand i've added an example of thread-safe version of spase function in the same pullrequest (#11534)
